libs/remote_storage/src/azure_blob.rs

```diff
@@ -664,3 +671,29 @@ where
         }
     }
 }
+
+/// A Stream wrapper that impls Sync even though the inner doesn't need to.
+struct SyncStream<T: Stream>(Mutex<Pin<Box<T>>>);
+
+impl<T: Stream> SyncStream<T> {
+    fn from_pin(inner: Pin<Box<T>>) -> Self {
+        SyncStream(Mutex::new(inner))
+    }
+}
+impl<T: Stream + Send> Stream for SyncStream<T> {
+    type Item = T::Item;
+    fn poll_next(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        let Ok(mut lock) = self.0.try_lock() else {
+            cx.waker().wake_by_ref();
+            return std::task::Poll::Pending;
+        };
+        let lock: Pin<&mut T> = Pin::as_mut(&mut lock);
+        Stream::poll_next(lock, cx)
+    }
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, None)
+    }
```

Comment:
Does the stream combination you ended up with give out something interesting or is this only "let's avoid locking again"?

Reply:
I have changed it to do locking and use the internal stream's `size_hint`.

Reply:
Ok... I don't think we actually use this. S3 sdk returns the *amount* of bytes which doesn't really fit this trait. Like if we knew the chunk size, we could give the same api should we start to use it.